asset-transfer-abac/chaincode-go/smart-contract/abac.go

```diff
@@ -0,0 +1,202 @@
+type Asset struct {
+	ID             string `json:"ID"`
+	Color          string `json:"color"`
+	Size           int    `json:"size"`
+	Owner          string `json:"owner"`
+	AppraisedValue int    `json:"appraisedValue"`
+}
+
+// CreateAsset issues a new asset to the world state with given details.
+func (s *SmartContract) CreateAsset(ctx contractapi.TransactionContextInterface, id string, color string, size int, appraisedValue int) error {
+
+	// Demonstrate the use of Attribute-Based Access Control (ABAC) by checking
+	// to see if the caller has the "abac.creator" attribute with a value of true;
+	// if not, return an error.
+	//
+	err := ctx.GetClientIdentity().AssertAttributeValue("abac.creator", "true")
+	if err != nil {
+		return fmt.Errorf("submitting client not authorized to create asset, does not have abac.creator role")
+	}
+
+	exists, err := s.AssetExists(ctx, id)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return fmt.Errorf("the asset %s already exists", id)
+	}
+
+	// Get ID of submitting client identity
+	clientID, err := s.GetSubmittingClientIdentity(ctx)
+	if err != nil {
+		return err
+	}
+
+	asset := Asset{
+		ID:             id,
+		Color:          color,
+		Size:           size,
+		Owner:          clientID,
+		AppraisedValue: appraisedValue,
+	}
+	assetJSON, err := json.Marshal(asset)
+	if err != nil {
+		return err
+	}
+
+	return ctx.GetStub().PutState(id, assetJSON)
+}
+
+// UpdateAsset updates an existing asset in the world state with provided parameters.
+func (s *SmartContract) UpdateAsset(ctx contractapi.TransactionContextInterface, id string, newColor string, newSize int, newValue int) error {
+
+	asset, err := s.ReadAsset(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	clientID, err := s.GetSubmittingClientIdentity(ctx)
+	if err != nil {
+		return err
+	}
+
+	if clientID != asset.Owner {
+		return fmt.Errorf("submitting client not authorized to update asset, does not own asset")
+	}
+
+	asset.Color = newColor
+	asset.Size = newSize
+	asset.AppraisedValue = newValue
+
+	assetJSON, err := json.Marshal(asset)
+	if err != nil {
+		return err
+	}
+
+	return ctx.GetStub().PutState(id, assetJSON)
+}
+
+// DeleteAsset deletes a given asset from the world state.
+func (s *SmartContract) DeleteAsset(ctx contractapi.TransactionContextInterface, id string) error {
+
+	asset, err := s.ReadAsset(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	clientID, err := s.GetSubmittingClientIdentity(ctx)
+	if err != nil {
+		return err
+	}
+
+	if clientID != asset.Owner {
+		return fmt.Errorf("submitting client not authorized to update asset, does not own asset")
+	}
+
+	return ctx.GetStub().DelState(id)
+}
+
+// TransferAsset updates the owner field of asset with given id in world state.
+func (s *SmartContract) TransferAsset(ctx contractapi.TransactionContextInterface, id string, newOwner string) error {
+
+	asset, err := s.ReadAsset(ctx, id)
+	if err != nil {
+		return err
+	}
+
```

Comment:
I assume you'd want to do the auth check `clientID != asset.Owner` here as well.

Reply:
Added.